weldx/measurement.py

```diff
@@ -37,71 +28,705 @@ class Signal:
-    ):  # pragma: no cover
-        """Add a new node to several containers.
-
-        This is a helper for the plot function.
-
-        Parameters
-        ----------
-        node :
-            Name of the new node
-        parent_node :
-            Name of the parent node
-        node_label :
-            Displayed name of the node
-        position :
-            Position of the node
-        container :
-            Tuple of containers that should be updated.
-
-        """
-        [graph, node_list, labels, positions] = container
-        graph.add_node(node)
-        node_list.append(node)
-        labels[node] = node_label
-        positions[node] = position
-        if parent_node is not None:
-            graph.add_edge(parent_node, node)
+    def _determine_output_signal_unit(
+        func: "MathematicalExpression", input_unit: str
+    ) -> str:
+        """Determine the unit of a transformations' output signal.
+
+        Parameters
+        ----------
+        func :
+            The function describing the transformation
+        input_unit :
+            The unit of the input signal
+
+        Returns
+        -------
+        str:
+            Unit of the transformations' output signal
+
+        """
+        from weldx import Q_
+
+        if func is not None:
+            variables = func.get_variable_names()
+            if len(variables) != 1:
+                raise ValueError("The provided function must have exactly 1 parameter")
+
+            try:
+                test_output = func.evaluate(**{variables[0]: Q_(1, input_unit)})
+            except Exception as e:
+                raise ValueError(
+                    "The provided function is incompatible with the input signals unit."
+                    f" \nThe test raised the following exception:\n{e}"
+                )
+            return str(test_output.units)
+
+        return input_unit
+
+    def _raise_if_data_exist(self, signal_source_name: str):
+        """Raise an error if the signal from the passed source already has data."""
+        self._raise_if_node_does_not_exist(signal_source_name)
+        if self._graph.nodes[signal_source_name]["signal"].data is not None:
+            raise KeyError(
+                "The measurement chain already contains data for "
+                f"'{signal_source_name}'."
+            )
+
+    def _raise_if_node_exist(self, node_id: str):
+        """Raise en error if the graph already contains a node with the passed id."""
+        if node_id in self._graph.nodes:
+            raise KeyError(
+                f"The internal graph already contains a node with the id {node_id}"
+            )
+
+    def _raise_if_node_does_not_exist(self, node: str):
+        """Raise a `KeyError` if the specified node does not exist."""
+        if node not in self._graph.nodes:
+            raise KeyError(f"No signal with source '{node}' found")
+
+    def add_signal_data(self, data: "TimeSeries", signal_source: str = None):
```

Comment:
```suggestion
    def add_signal_data(self, data: TimeSeries, signal_source: str = None):
```
I think this should work with `from __future__ import annotations`